src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,920 @@
+										 List *splitChildrenShardIntervalList,
+										 List *workersForPlacementList);
+static void InsertSplitChildrenShardMetadata(List *shardGroupSplitIntervalListList,
+											 List *workersForPlacementList);
+static void CreateForeignKeyConstraints(List *shardGroupSplitIntervalListList,
+										List *workersForPlacementList);
+static void TryDropSplitShardsOnFailure(List *shardGroupSplitIntervalListList,
+										List *workersForPlacementList);
+
+/* Customize error message strings based on operation type */
+static const char *const SplitOperationName[] =
+{
+	[SHARD_SPLIT_API] = "split",
+	[ISOLATE_TENANT_TO_NEW_SHARD] = "isolate",
+};
+static const char *const SplitTargetName[] =
+{
+	[SHARD_SPLIT_API] = "shard",
+	[ISOLATE_TENANT_TO_NEW_SHARD] = "tenant",
+};
+
+/* Function definitions */
+
+/*
+ * ErrorIfCannotSplitShard checks relation kind and invalid shards. It errors
+ * out if we are not able to split the given shard.
+ */
+void
+ErrorIfCannotSplitShard(SplitOperation splitOperation, ShardInterval *sourceShard)
+{
+	Oid relationId = sourceShard->relationId;
+	ListCell *colocatedTableCell = NULL;
+	ListCell *colocatedShardCell = NULL;
+
+	/* checks for table ownership and foreign tables */
+	List *colocatedTableList = ColocatedTableList(relationId);
+	foreach(colocatedTableCell, colocatedTableList)
+	{
+		Oid colocatedTableId = lfirst_oid(colocatedTableCell);
+
+		/* check that user has owner rights in all co-located tables */
+		EnsureTableOwner(colocatedTableId);
+
+		char relationKind = get_rel_relkind(colocatedTableId);
+		if (relationKind == RELKIND_FOREIGN_TABLE)
+		{
+			char *relationName = get_rel_name(colocatedTableId);
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot %s %s because \"%s\" is a "
+								   "foreign table",
+								   SplitOperationName[splitOperation],
+								   SplitTargetName[splitOperation],
+								   relationName),
+							errdetail("Splitting shards backed by foreign tables "
+									  "is not supported.")));
+		}
+
+		/*
+		 * At the moment, we do not support copying a shard if that shard's
+		 * relation is in a colocation group with a partitioned table or partition.
+		 */
+		if (PartitionedTable(colocatedTableId))
+		{
+			char *sourceRelationName = get_rel_name(relationId);
+			char *colocatedRelationName = get_rel_name(colocatedTableId);
+
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot %s of '%s', because it "
+								   "is a partitioned table",
+								   SplitOperationName[splitOperation],
+								   colocatedRelationName),
+							errdetail("In colocation group of '%s', a partitioned "
+									  "relation exists: '%s'. Citus does not support "
+									  "%s of partitioned tables.",
+									  sourceRelationName,
+									  colocatedRelationName,
+									  SplitOperationName[splitOperation])));
+		}
+	}
+
+	/* check shards with inactive placements */
```

Comment:
The notion of inactive placements is more or less deprecated. Instead of adding these checks, we could consider dropping the inactive placements as well in the drop phase.

Reply:
@marcocitus, Lets handle this in a follow up PR. This method is also shared with isolate_tenant_to_new_shards and I would like to keep the behavior as is for now.

Reply:
Not sure, how you guys handle pending items but I can file a bug / TODO on me if you want to explicitly track all these small items.

Reply:
Marking this as resolved. Have noted this down. Will follow up.